File: vrm_kit/glb.py
"""Low-level GLB (binary glTF) reader/writer.

VRM files are GLB files with VRM-specific extensions in the JSON chunk.
This module handles the binary container format directly, giving full
control over extension data without depending on any glTF library.
"""

from __future__ import annotations

import json
import struct
from pathlib import Path
from typing import Any

GLB_MAGIC = 0x46546C67  # 'glTF'
GLB_VERSION = 2
CHUNK_JSON = 0x4E4F534A
CHUNK_BIN = 0x004E4942


class GlbFile:
    __slots__ = ("json_data", "bin_data")

    def __init__(self, json_data: dict[str, Any], bin_data: bytes = b"") -> None:
        self.json_data = json_data
        self.bin_data = bin_data

    # ---- I/O ----

    @classmethod
    def load(cls, path: str | Path) -> GlbFile:
        data = Path(path).read_bytes()
        if len(data) < 12:
            raise ValueError("File too small to be GLB")

        magic, version, total_length = struct.unpack_from("<III", data, 0)
        if magic != GLB_MAGIC:
            raise ValueError(f"Not a GLB file (magic: {magic:#010x})")
        if version != GLB_VERSION:
            raise ValueError(f"Unsupported GLB version: {version}")

        offset = 12
        json_obj: dict[str, Any] | None = None
        bin_bytes = b""

        while offset < total_length and offset + 8 <= len(data):
            chunk_len, chunk_type = struct.unpack_from("<II", data, offset)
            offset += 8
            chunk_data = data[offset : offset + chunk_len]
            offset += chunk_len

            if chunk_type == CHUNK_JSON:
                json_obj = json.loads(chunk_data)
            elif chunk_type == CHUNK_BIN:
                bin_bytes = chunk_data

        if json_obj is None:
            raise ValueError("GLB file has no JSON chunk")

        return cls(json_obj, bin_bytes)

    def save(self, path: str | Path) -> None:
        json_bytes = json.dumps(
            self.json_data, ensure_ascii=False, separators=(",", ":")
        ).encode("utf-8")
        # Pad JSON to 4-byte boundary with spaces (per glTF spec)
        json_pad = (4 - len(json_bytes) % 4) % 4
        json_bytes += b" " * json_pad

        has_bin = bool(self.bin_data)
        bin_bytes = self.bin_data
        bin_pad = 0
        if has_bin:
            bin_pad = (4 - len(bin_bytes) % 4) % 4

        total = 12 + 8 + len(json_bytes)
        if has_bin:
            total += 8 + len(bin_bytes) + bin_pad

        out = bytearray()
        out += struct.pack("<III", GLB_MAGIC, GLB_VERSION, total)
        out += struct.pack("<II", len(json_bytes), CHUNK_JSON)
        out += json_bytes
        if has_bin:
            out += struct.pack("<II", len(bin_bytes) + bin_pad, CHUNK_BIN)
            out += bin_bytes
            out += b"\x00" * bin_pad

        Path(path).write_bytes(bytes(out))

    # ---- Image operations ----

    def extract_image(self, image_index: int) -> bytes:
        """Extract an image's raw data (PNG/JPEG bytes) from the binary chunk."""
        images = self.json_data.get("images", [])
        if image_index >= len(images):
            raise IndexError(f"Image index {image_index} out of range ({len(images)})")
        bv_index = images[image_index]["bufferView"]
        bv = self.json_data["bufferViews"][bv_index]
        offset = bv.get("byteOffset", 0)
        return self.bin_data[offset : offset + bv["byteLength"]]

    def replace_image(self, image_index: int, new_data: bytes) -> None:
        """Replace an image's data in the binary chunk, adjusting all offsets."""
        images = self.json_data.get("images", [])
        if image_index >= len(images):
            raise IndexError(f"Image index {image_index} out of range ({len(images)})")

        bv_index = images[image_index]["bufferView"]
        buffer_views = self.json_data["bufferViews"]
        bv = buffer_views[bv_index]
        old_offset = bv.get("byteOffset", 0)
        old_length = bv["byteLength"]
        size_diff = len(new_data) - old_length

        # Splice the binary chunk
        self.bin_data = (
            self.bin_data[:old_offset]
            + new_data
            + self.bin_data[old_offset + old_length :]
        )

        # Update this bufferView
        bv["byteLength"] = len(new_data)

        # Shift all subsequent bufferViews
        for other_bv in buffer_views:
            if other_bv is bv:
                continue
            other_offset = other_bv.get("byteOffset", 0)
            if other_offset > old_offset:
                other_bv["byteOffset"] = other_offset + size_diff

        # Update total buffer length
        buffers = self.json_data.get("buffers", [])
        if buffers:
            buffers[0]["byteLength"] = len(self.bin_data)

File: src-tauri/src/glb.rs
use serde_json::Value;
use std::path::Path;

const GLB_MAGIC: u32 = 0x46546C67;
const CHUNK_JSON: u32 = 0x4E4F534A;
const CHUNK_BIN: u32 = 0x004E4942;

#[derive(Clone)]
pub struct GlbFile {
    pub json: Value,
    pub bin: Vec<u8>,
}

impl GlbFile {
    pub fn load(path: &Path) -> Result<Self, String> {
        let data = std::fs::read(path).map_err(|e| format!("read: {e}"))?;
        Self::parse(&data)
    }

    pub fn parse(data: &[u8]) -> Result<Self, String> {
        if data.len() < 12 {
            return Err("too small".into());
        }
        let magic = u32::from_le_bytes(data[0..4].try_into().unwrap());
        if magic != GLB_MAGIC {
            return Err(format!("not GLB: {magic:#x}"));
        }
        let total = u32::from_le_bytes(data[8..12].try_into().unwrap()) as usize;

        let mut offset = 12usize;
        let mut json_val: Option<Value> = None;
        let mut bin_data: Vec<u8> = Vec::new();

        while offset + 8 <= total.min(data.len()) {
            let chunk_len = u32::from_le_bytes(data[offset..offset + 4].try_into().unwrap()) as usize;
            let chunk_type = u32::from_le_bytes(data[offset + 4..offset + 8].try_into().unwrap());
            offset += 8;
            let end = (offset + chunk_len).min(data.len());

            match chunk_type {
                CHUNK_JSON => {
                    json_val = Some(serde_json::from_slice(&data[offset..end]).map_err(|e| format!("json: {e}"))?);
                }
                CHUNK_BIN => {
                    bin_data = data[offset..end].to_vec();
                }
                _ => {}
            }
            offset = end;
        }

        Ok(Self {
            json: json_val.ok_or("no JSON chunk")?,
            bin: bin_data,
        })
    }

    pub fn save(&self, path: &Path) -> Result<(), String> {
        let json_bytes = serde_json::to_vec(&self.json).map_err(|e| format!("json: {e}"))?;
        let json_pad = (4 - json_bytes.len() % 4) % 4;
        let json_padded_len = json_bytes.len() + json_pad;

        let has_bin = !self.bin.is_empty();
        let bin_pad = if has_bin { (4 - self.bin.len() % 4) % 4 } else { 0 };

        let mut total = 12 + 8 + json_padded_len;
        if has_bin {
            total += 8 + self.bin.len() + bin_pad;
        }

        let mut out = Vec::with_capacity(total);
        out.extend_from_slice(&GLB_MAGIC.to_le_bytes());
        out.extend_from_slice(&2u32.to_le_bytes());
        out.extend_from_slice(&(total as u32).to_le_bytes());

        out.extend_from_slice(&(json_padded_len as u32).to_le_bytes());
        out.extend_from_slice(&CHUNK_JSON.to_le_bytes());
        out.extend_from_slice(&json_bytes);
        out.extend(std::iter::repeat(b' ').take(json_pad));

        if has_bin {
            out.extend_from_slice(&((self.bin.len() + bin_pad) as u32).to_le_bytes());
            out.extend_from_slice(&CHUNK_BIN.to_le_bytes());
            out.extend_from_slice(&self.bin);
            out.extend(std::iter::repeat(0u8).take(bin_pad));
        }

        std::fs::write(path, &out).map_err(|e| format!("write: {e}"))
    }

    pub fn extract_image(&self, idx: usize) -> Result<Vec<u8>, String> {
        let images = self.json["images"].as_array().ok_or("no images")?;
        let img = images.get(idx).ok_or("image index out of range")?;
        let bv_idx = img["bufferView"].as_u64().ok_or("no bufferView")? as usize;
        let bvs = self.json["bufferViews"].as_array().ok_or("no bufferViews")?;
        let bv = bvs.get(bv_idx).ok_or("bufferView out of range")?;
        let offset = bv["byteOffset"].as_u64().unwrap_or(0) as usize;
        let length = bv["byteLength"].as_u64().ok_or("no byteLength")? as usize;
        Ok(self.bin[offset..offset + length].to_vec())
    }

    pub fn replace_image(&mut self, idx: usize, new_data: &[u8]) -> Result<(), String> {
        let bv_idx = self.json["images"][idx]["bufferView"]
            .as_u64()
            .ok_or("invalid image")? as usize;

        let old_offset = self.json["bufferViews"][bv_idx]["byteOffset"]
            .as_u64()
            .unwrap_or(0) as usize;
        let old_length = self.json["bufferViews"][bv_idx]["byteLength"]
            .as_u64()
            .ok_or("no byteLength")? as usize;
        let size_diff = new_data.len() as i64 - old_length as i64;

        let mut new_bin = self.bin[..old_offset].to_vec();
        new_bin.extend_from_slice(new_data);
        new_bin.extend_from_slice(&self.bin[old_offset + old_length..]);
        self.bin = new_bin;

        self.json["bufferViews"][bv_idx]["byteLength"] = serde_json::json!(new_data.len());

        if let Some(bvs) = self.json["bufferViews"].as_array_mut() {
            for (i, bv) in bvs.iter_mut().enumerate() {
                if i == bv_idx {
                    continue;
                }
                let off = bv["byteOffset"].as_u64().unwrap_or(0) as i64;
                if off > old_offset as i64 {
                    bv["byteOffset"] = serde_json::json!((off + size_diff) as u64);
                }
            }
        }

        if let Some(buffers) = self.json["buffers"].as_array_mut() {
            if let Some(buf) = buffers.first_mut() {
                buf["byteLength"] = serde_json::json!(self.bin.len());
            }
        }

        Ok(())
    }
}
